### DISCUSSION TITLE
Why not use public key from private key file?

### DISCUSSION THREAD

**Discussion by d3x0r** [libssh2/libssh2#1330] [General]
I ended up digging into the public/private certificate features ssh uses, because I accidentally reversed public and private key being passed to the `_frommemory()` function, and the negotiation failed.  So I thought maybe libssh2 actually wanted a decoded pem file (instead of the base64 version native), which turns out it didn't... but I did find that the public and private keys are available in the single pem file anyway as part of the standard format for that file.   So that leads me to my question...

Is there some reason libssh2 always recomputes the public key from the private key instead of just using the public key provided?

**Comment by vszakats** [comment]
> Why not use public key from private key file?
> Is there some reason libssh2 always recomputes the public key from the private key instead of just using the public key provided?

Reading `libssh2_userauth_publickey_frommemory` in `src/userauth.c` it seems libssh2 accepts either a private key, in which case it extracts the public key from it automatically, or a public key directly, in which case it ignores the private key. In case both are provided, it uses the public key.

**Comment by d3x0r** [comment]
The following code is in `_libssh2_openssh_pem_parse_data`... it's where it gets the public key, ignores the result and then gets the public key.  The public key is never saved when given in the private key.

https://github.com/libssh2/libssh2/blob/master/src/pem.c#L487-L492

so like `gen_publickey_from_rsa_openssh_priv_data` is always called.
